server/src/main/java/io/spine/server/aggregate/UncommittedHistory.java

```diff
@@ -0,0 +1,206 @@
+ * modification, must retain the above copyright notice and the following
+ * disclaimer.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.server.aggregate;
+
+import com.google.common.collect.ImmutableList;
+import io.spine.core.Event;
+import io.spine.server.type.EventEnvelope;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Supplier;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * Uncommitted events and snapshots created for this aggregate during the dispatching.
+ *
+ * <p>Watches the how the events {@linkplain Aggregate#invokeApplier(EventEnvelope) are sent}
+ * to the {@link Aggregate} applier methods. Remembers all such events as uncommitted.
+ *
+ * <p>Once an aggregate is loaded from the storage, the {@code UncommittedHistory}
+ * {@linkplain #onAggregateRestored(AggregateHistory) remembers} the event count
+ * after the last snapshot.
+ *
+ * <p>If during the dispatching of events the number of events since the last snapshot exceeds
+ * the snapshot trigger, a snapshot is created and remembered as a part of uncommitted history.
+ *
+ * <p>In order to ignore the events fed to the aggregate when it's being loaded from the storage,
+ * the {@code UncommittedHistory}'s tracking is only {@linkplain #startTracking(int) activated}
+ * when the new and truly un-yet-committed events are dispatched to the applier methods.
+ * The tracking {@linkplain #stopTracking() stops} after all the new events have been played
+ * on the aggregate instance.
+ *
+ * @see Aggregate#apply(List, int) on activation and deactivation of event tracking
+ * @see Aggregate#replay(AggregateHistory) on supplying the history stats when loading aggregate
+ *         instances from the storage
+ */
+final class UncommittedHistory {
+
+    private final Supplier<Snapshot> makeSnapshot;
+    private final List<AggregateHistory> historySegments = new ArrayList<>();
+    private final List<Event> currentSegment = new ArrayList<>();
+
+    private int eventCountAfterLastSnapshot;
+    private @Nullable Integer snapshotTrigger = null;
```

Comment:
The reason is that the `UncommittedHistory` is a part of an `Aggregate` instance. But the value of a snapshot trigger is a knowledge outer to the `Aggregate`. It is brought by a corresponding repository at a time of applying the events inside the `AggregateEndpoint`. In turn, the repository also has this value mutable, so that end-users are able to re-define it.

Therefore it's not possible to instantiate an Aggregate instance—and thus its `UncommittedHistory`—with the snapshot trigger set right away.